Set up build for vendored package

oandregal opened the pull request:
Follow-up to https://github.com/Automattic/wp-calypso/pull/102276

## Proposed Changes

Sets up the build processes for `packages/dataviews` vendored copy.

## Tasks

- [x] Set it up as part of the yarn workspace.
  - [x] `clean`.
  - [x] build `packages/dataviews` upon running `yarn` (`prepare` command that calls `build`).
  - [x] remove exclusion from `wp-calypso/package.json`.
- [x] Compiled code (commands for the package):
  - [x] Output in `build/`: CJS.
  - [x] Output in `build-module/`: ESM.
  - [x] Output in `build-types/`: typescript types.
  - [x] Output in `build-style/`. `build:style` command.
    - [x] There are small differences from what's published in npm, we need sync from upstream to bring those changes in. For example, https://github.com/WordPress/gutenberg/commit/87b3af7957de01403da841a7a3945599d3263742 and https://github.com/WordPress/gutenberg/commit/10ae0d7d76bab76eff590985e13ab9c824b12d5a). https://github.com/Automattic/wp-calypso/pull/102704
    - [x] Pass base-styles differently. 
    - [x] `--wp-admin-theme-color-darker-10`: it's set to `rgb(0, 107, 160.5)` but it should be set to `#006ba1`.
  - [x] Output in `build-wp/`. Build WordPress bundle, `build:wp` command.
    - [x] There're some differences in `PopoverSlot` due to this PR using a different version of `@wordpress/components`. It's using 29.7.0 while the latest published one (which `@wordpress/dataviews` uses) is 29.8.0. This will be addressed in https://github.com/Automattic/wp-calypso/pull/102614
- [x] npm package preparation: review README, `package.json`, CHANGELOG, LICENSE, etc.

## Testing Instructions

- **Compare the output of styles and JavaScript** to what's available in the npm registry as `@wordpress/dataviews` 4.18.0.

| `@wordpress/dataviews` | This PR |  | 
| --- | --- | --- |
| `build-style/style.css` | `dist/style/style.css` | They should be exactly the same. | 
| `build-style/style-rtl.css` | `dist/style/style-rtl.css` | They should be exactly the same. |
| `build-wp/index.js` | `dist/wp/index.js` | The expectation is that there are no changes, other than paths in comments (e.g.: from `// ../../node_modules/@wordpress/components/...` to `// ../components/.` and a `flexShrink` value that changes from string to integer. | 

- **Test this PR with a small app.**

I've tested it by building a small app at ([dv-test](https://github.com/oandregal/dv-test)).

```sh
git clone git@github.com:oandregal/dv-test.git
npm install
npm start
```

It'll open a new browser tab at http://localhost:3000/ and you can interact with the data. The app bundles the latest changes in this PR as a tgz (`automattic-dataviews-0.1.0.tgz`), so it simulates it's using the `@automattic/dataviews` package.

oandregal committed:
Make packages/dataviews part of yarn workspace

oandregal committed:
Ignore build folders

oandregal committed:
Add build:style command

oandregal committed:
Add react-dom as a peer dep (components)

oandregal committed:
Add clean command to packages/dataviews

oandregal committed:
Add calypso:src as per docs/monorepo.md

oandregal committed:
Set up tsconfig for packages/dataviews

oandregal committed:
Add build:js command

oandregal committed:
Update build:style command

oandregal committed:
Add build and prepare commands

oandregal commented on packages/dataviews/src/style.scss:
```diff
@@ -1,3 +1,11 @@
+@import "../../node_modules/@wordpress/base-styles/colors";
```
TODO: fix this.

oandregal commented on packages/dataviews/package.json:
```diff
@@ -20,6 +20,7 @@
 	},
 	"main": "build/index.js",
 	"module": "build-module/index.js",
+	"calypso:src": "src/index.ts",
```
I don't know what this property is for, but `docs/monorepo` [mention it as required](https://github.com/Automattic/wp-calypso/blob/trunk/docs/monorepo.md#:~:text=Your%20package.json%20can%20have%20any%20of%20the%20normal%20properties%20but%20at%20a%20minimum%20should%20contain%20main%2C%20module%2C%20calypso%3Asrc%20and%20sideEffects.).

jsnajdr replied:
This is for the Calypso build (the main app) to be able to consume source files in packages without first running the build step -- the one that creates the `build` or `dist` folders.

Calypso webpack build can work directly with the source (TS) files, it has all the TS/Babel transpilation set up. The `dist`/`build` directories are only for consumers that get the package from NPM.

Please add the same `calypso:src` field to the `exports`/`"."` definition a few lines below. We'll soon migrate the Calypso tooling setup to the more modern `exports` definition, in another PR.

oandregal replied:
I pushed the suggested change.

oandregal committed:
Declare base-styles load path in includePaths

oandregal commented on packages/dataviews/build-style.js:
```diff
@@ -0,0 +1,138 @@
+
+	const builtSass = await renderSass( {
+		file,
+		includePaths: [ '../../node_modules/@wordpress/base-styles' ],
```
[includePaths](https://sass-lang.com/documentation/js-api/interfaces/legacystringoptions/#includePaths) points to a place that can resolve imports.

Upstream [includes a path](https://github.com/WordPress/gutenberg/blob/trunk/bin/packages/build-worker.js#L125) that is not _installed_ (`packages/base-styles`) but we can't do that and need to point to the installed package here. This package is installed at the root, not in `packages/dataviews/node_modules`. Is there a better way to do this?

oandregal replied:
Alternatively, we could have a different entry point than `src/style.scss` that imports the `importList` files directly (e.g.: `@import "@wordpress/base-styles/colors";`). Unless there's a strong objection against this approach, I'd prefer to follow upstream as much as possible.

jsnajdr replied:
This `includePaths` looks fine, the only change I'd recommend is to delegate the actual resolving to Node:
```js
// finds the package in the ../../node_modules dir
const path = require.resolve( '@wordpress/base-styles' );
// converts the resolved index.js filename to a package root (where the package.json is)
const packagePath = require( 'pkg-dir' ).sync( path );
```
The same lookup is already happening in the `client/webpack.config.js` file.

oandregal replied:
👍 pushed that change

oandregal committed:
Remove unnecessary code

oandregal committed:
Update yarn.lock after rebase

oandregal committed:
Revert "Set up tsconfig for packages/dataviews"

This reverts commit cce711fd05a714b980db548cfb239f531c351fad.

oandregal committed:
Update .gitignore

jsnajdr commented on packages/dataviews/.gitignore:
```diff
@@ -0,0 +1,6 @@
+build-style/
+build-types/
+build-wp/
+node_modules/
```
A better home for these ignores would be the root `.gitignore` file. There it's common to ignore also package-specific directories like `/packages/dataviews/build`.

oandregal replied:
👍 pushed these changes

jsnajdr commented on packages/dataviews/babel.config.js:
```diff
@@ -0,0 +1,8 @@
+		presets: [ '@wordpress/babel-preset-default' ],
+		plugins: [ '@emotion/babel-plugin', 'babel-plugin-inline-json-import' ],
+	};
+};
```
It would be great if we could replace this Babel build (copy&pasted from Gutenberg) with TypeScript. TypeScript is able to do all the transpilation we need here, and also can generate CJS or ESM modules. We already use it to build all the other Calypso `/packages` -- just look at the `tsconfig.json` and `tsconfig-cjs.json` files and how they are invoked in the `build` scripts.

The Gutenberg Babel config has two additional plugins:
- Emotion: the DataViews package doesn't use it, does it?
- inline JSON import: TypeScript cannot do it (you need a bundler anyway to do it really 100% correct) but again: does the DataViews package do JSON imports? Probably not. In Gutenberg, JSON imports are typical for blocks.

All in all, TypeScript as configured in the Calypso repo should be able to build the package just fine.

oandregal replied:
It sounds sensible, I'll look into that.

jsnajdr replied:
For reference, I'm working on something very similar -- setting up a dual ESM/CJS build with TypeScript -- in a Jetpack package in this PR: https://github.com/Automattic/jetpack/pull/43106

oandregal replied:
I've pushed an initial `tsconfig.json`, need to fix some issues before it's ready.

jsnajdr commented on packages/dataviews/build-style.js:
```diff
@@ -0,0 +1,138 @@
+ *
+ * @type {Function}
+ */
+const makeDir = promisify( fs.mkdir );
```
This is interesting, the code apparently predates the `fs/promises` module 🙂

oandregal replied:
Yeah, this has been extracted from [build-workers](https://github.com/WordPress/gutenberg/blob/trunk/bin/packages/build-worker.js) as it is. Most of it has been written a few years ago, and I suppose there could be simpler ways to do the same today. I didn't think it was worth refactoring this, as it's convenient to be able to compare with upstream.

oandregal committed:
Update package lookup

oandregal committed:
Move ignores to top-level

oandregal committed:
Use typescript instead of babel to compile TS to JS

oandregal committed:
Revert changes to trunk

oandregal committed:
Delete unused babel.config

oandregal commented on packages/dataviews/tsconfig.json:
```diff
@@ -1,25 +1,43 @@
 {
 	"$schema": "https://json.schemastore.org/tsconfig.json",
-	"extends": "../../tsconfig.base.json",
```
I've copied the `tsconfig.base.json` from upstream into here.

oandregal committed:
Fix TypeScript issue in dataviews-item-actions

oandregal committed:
Remove unnecessary typeRoots

oandregal committed:
Add jest types

oandregal committed:
Add testing library types types

oandregal committed:
Add tsconfig.cjs.json

oandregal committed:
Use tsconfig-csj.json

oandregal committed:
Use tsc for types and js

oandregal committed:
Update .gitignore

oandregal committed:
Sort compiler options alphabetically

oandregal committed:
Add sourcemaps

oandregal committed:
Transform JSX files into JS and use _jsx to render them

oandregal committed:
rename skipDefaultLibCheck to skipLibCheck as per the docs

oandregal committed:
Fix typo

oandregal commented:
@jsnajdr I've configured this to use `tsconfig.json`, as suggested. I've configured it to use the same output folders as `@wordpress/dataviews`, though the actual files are going to be different given they use different compilers — I suppose that's fine for consumers.

I've reviewed all the options for the TSC, and they look fine to me. I've taken the upstream tsconfig as base and only modified things that I needed, so it's a bit different from what the base calypso tsconfig.

oandregal committed:
Tell stylelint to skip code in build-style folder

oandregal committed:
Skip build folders for jest tests

oandregal committed:
Update yarn.lock

oandregal committed:
Properly scape build folders

oandregal committed:
Add calypso:src to exports object

jsnajdr commented on packages/dataviews/package.json:
```diff
@@ -20,10 +20,12 @@
 		".": {
 			"types": "./build-types/index.d.ts",
 			"import": "./build-module/index.js",
+			"calypso:src": "./src/index.ts",
```
The `calypso:src` condition should be the first one, before `types` and `import`. Order matters here, the resolver in webpack/TypeScript/Jest will use the first condition that matches. And we always want it to be the `calypso:src` one.

oandregal replied:
pushed

jsnajdr commented on packages/dataviews/src/components/dataviews-item-actions/index.tsx:
```diff
@@ -192,11 +192,12 @@ export default function ItemActions< Item >( {
 			className="dataviews-item-actions"
 			style={ {
-				flexShrink: '0',
+				flexShrink: 0, // TODO: why is this change necessary?
```
Yes, why is this necessary and who (which tool) forced you to do it? 🙂

oandregal replied:
I investigated this now and think this is fine.

What I know:

- Gutenberg ends up using `csstype 3.1.2`, that defines the property as:

```typescript
export type FlexShrink = Globals | (number & {}) | (string & {});
```

Both number or string types are valid. This can be seen in the `packages/dataviews` code, where there's another [usage](https://github.com/Automattic/wp-calypso/blob/eafba877442c5d816ae4f9286b38fd22a59c5f0a/packages/dataviews/src/components/dataviews/index.tsx#L161) of `flexShrink` that is a number, not a string. That's why Gutenberg didn't report anything.

- Calypso ends up using `csstype 3.0.7`, which defines it in a more restrictive way:

```typescript
export type FlexShrink = Globals | (number & {});
```

that's why it reports it.

`csstype` is a dependency of `@types/react`. Gutenberg depends on the 18.3.1 version that depends on `"csstype": "^3.0.2"`. Calypso depends on the 18.3.18 version that depends on `"csstype": "^3.0.2"`. So, there must be another project in Gutenberg that requires a newer version of `csstype` (which is more permissive).

I don't think it's worth investigating this prop type change further, so I went ahead and removed the TODO. This thread and the commit comment will serve as future reference.

jsnajdr commented on packages/dataviews/tsconfig.json:
```diff
@@ -1,25 +1,31 @@
 {
 	"$schema": "https://json.schemastore.org/tsconfig.json",
-	"extends": "../../tsconfig.base.json",
```
You can try to extend from the Calypso TS config base:
```
"extends": "@automattic/calypso-typescript-config/ts-package.json"
```
That should make the list of `compilerOptions` much smaller.

oandregal replied:
I considered this, but had second thoughts. Upstream created the JS output using a babel script, so we're already changing how the output is produced. So, thinking this further, I suppose embracing calypso rules would be ok, and I've gone ahead and did this.

For what is worth:

- this enabled some extra flags that aren't set upstream: `disableReferencedProjectLoad`, `importHelpers`, `noEmitHelpers`, `forceConsistentCasingInFileNames`. I checked what they do, and they seem sensible.
- `esModuleInterop`: I've let it set to `false` because that's what upstream does. It was introduced [here](https://github.com/WordPress/gutenberg/pull/18942/files#diff-0b280a445be167f54cd916c0718c952b8e0d4ca9621903d05eec25efd4c6ed6dR30). I'm ccing @sirreal in case he has any thoughts, though I reckon it's been a few years since that change, and we didn't use TSC upstream to compile this code into CJS, as far as I'm aware.

sirreal replied:
Seems reasonable. I think we could have kept more of the extended config instead of overriding (like `es2022` instead of `esnext` for several settings) to keep more packages with the same configuration.

oandregal committed:
set calypso:src as the first option

oandregal committed:
Extend calypso tsconfig

oandregal committed:
Remove TODO

I investigated this now and think this is fine.

What I know:

- Gutenberg ends up using `csstype 3.1.2`, that defines the property as:

```typescript
export type FlexShrink = Globals | (number & {}) | (string & {});
```

Both number of string types are valid. This can be seen in the `packages/dataviews` code, where there's another [usage](https://github.com/Automattic/wp-calypso/blob/eafba877442c5d816ae4f9286b38fd22a59c5f0a/packages/dataviews/src/components/dataviews/index.tsx#L161) of `flexShrink` that is a number, not a string. That's why Gutenberg didn't report anything.

- Calypso ends up using `csstype 3.0.7`, which defines it in a more restrictive way:

```typescript
export type FlexShrink = Globals | (number & {});
```

that's why it reports it.

`csstype` is a dependency of `@types/react`. Gutenberg depends on the 18.3.1 version that depends on `"csstype": "^3.0.2"`. Calypso depends on the 18.3.18 version that depends on `"csstype": "^3.0.2"`. So, there must be another project in Gutenberg that requires a newer version of `csstype` (which is more permissive).

I don't think it's worth investigating this prop type change further, so I went ahead and removed the TODO.

oandregal committed:
Remove required prop

oandregal committed:
Simplify CSS imports

oandregal commented:
> **Why the color value for `--wp-admin-theme-color-darker-10` is `rgb(0, 107, 160.5)` instead of `#006ba1`? Does it matter?**
>
> I've been looking at these questions today and the answers are: 1) I don't know, 2) I think it doesn't and we should probably move on.

This is the full context:

The `build:style` command generates the stylesheet under `build-style/style.css` (also the RTL version), and it bundles some CSS Custom Properties defined in the `@wordpress/base-styles` package.

Input ([mixins](https://github.com/WordPress/gutenberg/blob/b23709686bea6c64b33fd476437b0d5cfd68a382/packages/base-styles/_mixins.scss#L506-L523) and [default-custom-properties](https://github.com/WordPress/gutenberg/blob/b23709686bea6c64b33fd476437b0d5cfd68a382/packages/base-styles/_default-custom-properties.scss#L5)):

```scss
@mixin admin-scheme($color-primary) {
	--wp-admin-theme-color-darker-10: #{darken($color-primary, 5%)};
	--wp-admin-theme-color-darker-10--rgb: #{hex-to-rgb(darken($color-primary, 5%))};
	--wp-admin-theme-color-darker-20: #{darken($color-primary, 10%)};
	--wp-admin-theme-color-darker-20--rgb: #{hex-to-rgb(darken($color-primary, 10%))};
}

:root {
  @include admin-scheme(#007cba);
}
```

Output:

```css
:root {
  --wp-admin-theme-color-darker-10: rgb(0, 107, 160.5);
  --wp-admin-theme-color-darker-10--rgb: 0, 107, 161;
  --wp-admin-theme-color-darker-20: #005a87;
  --wp-admin-theme-color-darker-20--rgb: 0, 90, 135;
}
```

Note the color value for `--wp-admin-theme-color-darker-10`. In the published `@wordpress/dataviews` package ([4.18.0 version](https://www.npmjs.com/package/@wordpress/dataviews/v/4.18.0), shares the same code as this PR), the output is slightly different, it has been compiled to the HEX value:

```css
--wp-admin-theme-color-darker-10: #006ba1;
```

That's the only difference in the whole file.

Considering that HEX values represent the same color space as RGB and that `#006ba1` translates to `rgb(0, 107, 161)` I think it's fine. I investigated why this happened, and I don't have an answer. If designers confirm this is ok of a change, I'll move on.

For reference, this is what I've checked so far:

- Both Gutenberg (`buildCSS` in [build-worker](https://github.com/WordPress/gutenberg/blob/11adb45dcf87204a17604c7a5d67bb3d36b4508e/bin/packages/build-worker.js#L123)) and Calypso (`buildCSS` function in `build-style` in this PR that has been copied from Gutenberg) use the same process and content:

```javascript
	const builtSass = await renderSass( {
		file,
		includePaths: [ findPackagePath( '@wordpress/base-styles' ) ],
		data: ''.concat( '@use "sass:math";', importLists, contents ),
	} );
```

- The `sass` version in wp-calypso and gutenberg is the same (1.54.0).
- In Gutenberg, if I execute the `npm run build:packages` command (which in turn calls build-worker), I'm getting the HEX color (same as the `@wordpress/dataviews` package).
- I tested adding a new content that was essentially `const test = ':root{ --my-color: #{darken(#007cba, 5%)};}';` and it outputs `:root{--my-color: rgb(0, 107, 160.5);}`.
- As per the docs, it doesn't seem that the [`darken` function](https://sass-lang.com/documentation/modules/color/#darken) can take a config/color space/etc. via environmental variable or similar. Checked [the code](https://github.com/sass/dart-sass/blob/3b8afa2b3e97174499af031d2cfceb47fb84b42a/lib/src/functions/color.dart#L210), and it doesn't look like it can. I haven't found any evidence that Gutenberg is configuring this differently.
- I've also reviewed older versions of `@wordpress/dataviews` package published in the registry and all of them have the HEX color.

I ran out of ideas to try today, and I'm a bit puzzled as to why I'm getting a different color value. I'm sharing this with some folks in case they have more ideas to address it. If nothing comes to mind, and it's confirmed innocuous, I'll leave it as it is for now and will move on to other things.

oandregal committed:
Pin the sass version to 1.54.0

oandregal commented:
> The sass version in wp-calypso and gutenberg is the same (1.54.0).

This is true. However, the `packages/dataviews/package.json` didn't pin the version, like the top-level `package.json` does. As a result, there were two copies of the sass compiler. 1.54.0 at the top-level and 1.87.0 at `packages/dataviews`. Apparently, the darken function behaves differently between the two.

I pushed a fix and the stylesheet is now the same.

oandregal committed:
Add LICENSE file

oandregal committed:
Update README

oandregal committed:
Set package.json package to public

oandregal committed:
Pin typescript to the version of calypso package.json

oandregal committed:
Add CHANGELOG for calypso

jsnajdr committed:
Switch build folders to dist

oandregal commented:
I've tested this PR in a small app I've built at ([dv-test](https://github.com/oandregal/dv-test)).

```sh
git clone git@github.com:oandregal/dv-test.git
npm install
npm start
```

It'll open a new browser tab at http://localhost:3000/ and you can interact with the data.

The app bundles this PR as a tgz (`automattic-dataviews-0.1.0.tgz`), so it simulates it's using the `@automattic/dataviews` package.

At first, I tried using local paths, but they end up being problematic with React hooks. For hooks to work, the react import from your application code needs to resolve to the same module as the react import from inside the react-dom package: hooks rely on a global centralized registry, if there are two React versions, there are two different registries that don't talk to each other. This is easier said than done: [docs](https://react.dev/warnings/invalid-hook-call-warning#mismatching-versions-of-react-and-react-dom), [other experiences](https://github.com/facebook/react/issues/13991). After some trying to make them work with local paths, I've switched to bundle the tgz.

jsnajdr reviewed (approved):
Looks good :+1:
